ui/src/pages/backoffice/DocumentRequest/DocumentRequestDetails/components/DocumentRequestSteps/DocumentRequestSteps.js

```diff
@@ -0,0 +1,131 @@
+    const title = _get(this.props, 'data.metadata.document.title');
+    return docPid ? (
+      <Segment basic>
+        <Grid columns={2}>
+          <Grid.Column width={12}>
+            <Header>
+              Attached Document{' '}
+              <Link to={BackOfficeRoutes.documentDetailsFor(docPid)}>
+                {title}
+              </Link>
+            </Header>
+          </Grid.Column>
+          <Grid.Column width={4}>
+            <Button
+              fluid
+              name="change-document"
+              onClick={this.onChangeDocument}
+              content={'Change document'}
+            />
+          </Grid.Column>
+        </Grid>
+      </Segment>
+    ) : (
+      <Message info>
+        <Message.Header>A document is required!</Message.Header>
+        <p>You can create one or search in existing documents</p>
+      </Message>
+    );
+  }
+
+  renderSteps(hasDocument, hasProvider) {
+    return (
+      <Step.Group attached="top" fluid>
+        <Step active={!hasDocument}>
+          <DocumentIcon />
+          <Step.Content>
+            <Step.Title>Attach Document</Step.Title>
+            <Step.Description>
+              Choose a document for your request
+            </Step.Description>
+          </Step.Content>
+        </Step>
+
+        <Step active={hasDocument && !hasProvider} disabled={!hasDocument}>
+          <Icon name="truck" />
+          <Step.Content>
+            <Step.Title>Select Provider</Step.Title>
+            <Step.Description>
+              Acquire or borrow from another Library
+            </Step.Description>
+          </Step.Content>
+        </Step>
+
+        <Step disabled>
+          <Icon name="check" />
+          <Step.Content>
+            <Step.Title>Confirm Request</Step.Title>
+          </Step.Content>
+        </Step>
+      </Step.Group>
+    );
+  }
+
+  renderStepsContent(hasDocument, hasProvider) {
+    const { data } = this.props;
+    if (!hasDocument) {
+      return (
+        <DocumentAttachStep
+          data={data}
+          fetchDocumentRequestDetails={this.props.fetchDocumentRequestDetails}
+        />
+      );
+    } else if (hasDocument && !hasProvider) {
+      return <DocumentProvidersStep hasDocument={hasDocument} />;
+    }
+    return <Header>Document Request information</Header>;
+  }
+
+  onChangeDocument = async () => {
+    const { pid } = this.props.data;
+    const resp = await documentRequestApi.pending(pid);
+    if (resp.status === 202) {
+      this.props.fetchDocumentRequestDetails(pid);
+    }
+  };
+
+  render() {
+    const { document_pid: docPid, state } = this.props.data.metadata;
+    const hasDocument = docPid ? true : false;
+    // TODO: when we integrated ill_pid or acq_pid
+    const hasProvider = false;
+    return state !== 'REJECTED' ? (
+      <>
+        {this.documentInfo()}
+        {this.renderSteps(hasDocument, hasProvider)}
+        <Segment>{this.renderStepsContent(hasDocument, hasProvider)}</Segment>
```

Comment:
Right now the calculation to figure out the current "step state" happens independently in both `renderSteps` and `renderStepsContent`. This might be a bit messy to keep these synchronized in case there are changes so maybe we could have that calculation happen once and store the "step state" and pass to these functions. What do you think?

Reply:
I totally agree, there should be one source of truth instead of applying fragile logic everywhere. I had some issues rendering while also applying updates to the state.